File: utils/md_generator.py
from common import *
from function_wrapper import FunctionWrapper

def generate_md_text(version="reborn"):
    def generate_md_pre_contents(version="reborn"):
        return f'''---
title: Pokemon {version.capitalize()} Walkthrough
---

<p id="title-text">Pokemon {version.capitalize()} Walkthrough </p>

'''
  
    def generate_toc_contents():
        toc = ""
        for chapter_type, total_chapters in SECTIONS[version]:
            for chapter_num in range(1, total_chapters + 1):
                raw_md = load_chapter_md(version, chapter_type, chapter_num)
                for line in [line for line in raw_md.split('\n') if line.startswith('#')]:
                    indents = len(line) - len(line.lstrip('#')) - 1
                    next_line = f"""{"  " * indents}- [{line.lstrip('#')[1:]}](#{
                        line.lstrip('#')[1:]
                        .lower()
                        .translate(str.maketrans('', '', string.punctuation.replace('-', '')))
                        .replace(' ', '-')
                    })\n"""
                    toc += next_line
        return toc + '\n\n'
   
    def generate_md_post_contents():
        # TODO
        return ''
    
    def generate_chapter_contents(type='main', num=1):
        raw_md =  load_chapter_md(version, type, num) 

        # Store chapter text as an array of lines - join them at the end
        res = []
        for line in raw_md.split('\n'):
            if line == '' or line[0] != '!':
                res.append(line)
            elif line[0] == '!':
                #Function Wrapper class does the magic of taking a line 
                # beginning with ! and transforming it into a dynamic output:
                # taking a shortened function name, arguments, and globals
                function_result = function_wrapper.evaluate_function_from_string(line)
                res.append(function_result)
        return '\n'.join(res)
    
    function_wrapper = FunctionWrapper(version)

    res = ''
    res += generate_md_pre_contents()
    res += generate_toc_contents()
    for chapter_type, total_chapters in SECTIONS[version]:
        for chapter_num in range(1, total_chapters + 1):
            res += generate_chapter_contents(chapter_type, chapter_num)
    res += generate_md_post_contents()

    return res


if __name__ == "__main__":
    md = generate_md_text()
    # print(md)

File: utils_ruby/md_generator.rb
require_relative 'common'
require_relative 'function_wrapper'

def generate_md_text(game = "reborn")
  func_wrapper = FunctionWrapper.new(game)

  def generate_md_pre_contents(game = "reborn")
    <<~PRE_CONTENTS
      ---
      title: Pokémon #{game.capitalize} Walkthrough
      ---

      <p id="title-text">Pokémon #{game.capitalize} Walkthrough </p>

    PRE_CONTENTS
  end
  
  def generate_toc_contents(game)
    toc = ""
    SECTIONS[game].each do |chapter_type, total_chapters|
      (1..total_chapters).each do |chapter_num|
        raw_md = load_chapter_md(game, chapter_type, chapter_num)
        raw_md.each_line do |line|
          if line.start_with?('#')
            indents = line[/^#+/].length - 1
            title = line.strip[indents + 1..].strip  # Remove the leading '#' and any extra spaces
            anchor_link = title.downcase.gsub(/[^a-z0-9 -]/, '').tr(' ', '-')  # Convert title to lowercase, remove non-alphanumeric characters except spaces, and replace spaces with dashes
            toc += "#{'  ' * indents}- [#{title}](##{anchor_link})\n"
          end
        end
      end
    end
    toc + "\n"
  end

  def generate_md_post_contents
    # TODO
    ''
  end
  
  def generate_chapter_contents(game, type, num, func_wrapper)
    raw_md = load_chapter_md(game, type, num) 

    # Store chapter text as an array of lines - join them at the end
    res = []
    raw_md.each_line do |line|
      if line.strip.empty? || line[0] != '!'
        res << line
      elsif line[0] == '!'
        # Function Wrapper class does the magic of taking a line 
        # beginning with ! and transforming it into a dynamic output:
        # taking a shortened function name, arguments, and globals
        function_result = func_wrapper.evaluate_function_from_string(line)
        res << function_result
      end
    end
    return res.join
  end

  res = ''
  res += generate_md_pre_contents(game)
  res += generate_toc_contents(game)
  SECTIONS[game].each do |chapter_type, total_chapters|
    (1..total_chapters).each do |chapter_num|
      res += generate_chapter_contents(game, chapter_type, chapter_num, func_wrapper)
      res += "\n"
    end
  end
  res += generate_md_post_contents
  res.strip
  res
end